accepted/new-api.md

```diff
@@ -0,0 +1,1495 @@
+documentation.
+
+All other existing methods in the `wasmtime` API are updated to take a "context"
+argument. To assist with the ergonomics of this argument new traits are added as
+well to the above two types:
+
+```rust
+trait AsContext {
+    type Data;
+    fn as_context(&self) -> StoreContext<'_, Self::Data>;
+}
+
+trait AsContextMut: AsContext {
+    fn as_context_mut(&mut self) -> StoreContextMut<'_, Self::Data>;
+}
+
+// forward `&T => T`
+impl<'a, T: AsContext + ?Sized> AsContext for &'a T {
+    type Data = T::Data;
+    fn as_context(&self) -> StoreContext<'_, T::Data> {
+        T::as_context(self)
+    }
+}
+
+// forward `&mut T => T`
+impl<'a, T: AsContext + ?Sized> AsContext for &'a mut T {
+    type Data = T::Data;
+    fn as_context(&self) -> StoreContext<'_, T::Data> {
+        T::as_context(self)
+    }
+}
+
+// forward `&mut T => T`
+impl<'a, T: AsContextMut + ?Sized> AsContextMut for &'a mut T {
+    fn as_context_mut(&mut self) -> StoreContextMut<'_, T::Data> {
+        T::as_context_mut(self)
+    }
+}
+
+// Implementations for store-related and context types
+impl<T> AsContext for Store<T> {
+    type Data = T;
+    fn as_context(&self) -> StoreContext<'_, T> { self.context() }
+}
+impl<T> AsContextMut for Store<T> {
+    fn as_context_mut(&mut self) -> StoreContextMut<'_, T> { self.context_mut() }
+}
+impl<T> AsContext for StoreContext<'_, T> {
+    type Data = T;
+    fn as_context(&self) -> StoreContext<'_, T> { self.as_ref() }
+}
+impl<T> AsContext for StoreContextMut<'_, T> {
+    type Data = T;
+    fn as_context(&self) -> StoreContext<'_, T> { self.as_ref() }
+}
+impl<T> AsContextMut for StoreContextMut<'_, T> {
+    fn as_context_mut(&mut self) -> StoreContextMut<'_, T> { self.as_mut() }
+}
+```
+
+with these trait definitions you'll see parameters of one of two form in APIs
+below:
+
+1. `cx: impl AsContext` - this is for APIs which are safe to be called
+   concurrently from many threads. For example learning about type information
+   or something like `Global::get`.
+2. `cx: impl AsContextMut` - this is the more common case for APIs which require
+   exclusive access to call (therefore can't be called concurrently). This is
+   used for cases which could mutate the store such as `Func::new` or
+   `Func::call`.
+
+The goal of these traits and these implementations is to make it as ergonomic as
+we can to pass whatever kind of context you have into a function requiring a
+context and it'll work.
+
+### `Func`
+
+Using the context types defined above the `Func` API will look similar to what
+it is today except that methods now take a context parameter. There are two
+subtle but impactful differences in this API worth calling out:
+
+* All closures used to create a `Func` are required to be `Send + Sync`. This is
+  because the types are stored in `Store` and we want `Store` to be `Send +
+  Sync`. You still only get `Fn`, though, instead of `FnMut`. Mutability of host
```

Comment:
Above it is suggested that having the `Store` be `Send + Sync` is optional and depends on `T` being `Send + Sync` too; so is this a strong requirement, even when the embedder doesn't need a `Send + Sync` `Store`?

Reply:
That's correct, regardless of whether the embedder needs `Send` and `Sync` host functions will *always* be required to be `Send` and `Sync`. This is thought to be ok, though, because most host state should be stored in `T` rather than in the funtion itself, and it's expected that the majority of host functions are simply 0-sized closures that are just code and no data.